In [100]:
# import all necessory packages
import pandas as pd
import numpy as np

In [101]:
# Load data from csv file
data = pd.read_csv("mercedesbenz.csv", usecols=["X1","X2","X3","X4","X5","X6"])

In [102]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [103]:
# check how many unique labels have in each column
for col in data.columns:
  print(col, ': ', len(data[col].unique()), 'labels')

X1 :  27 labels
X2 :  44 labels
X3 :  7 labels
X4 :  4 labels
X5 :  29 labels
X6 :  12 labels


In [104]:
# examine the number of columns after one hot coding with pandas in build method
pd.get_dummies(data, drop_first=True).shape

(4209, 117)

##### Here, we are getting 117 new variables from 6 initial categorical variables. Instaed, We can select top 10 most frequent categories from each variable.

In [105]:
# Let's find top 10 categories from 'X2' variable
data['X1'].value_counts().sort_values(ascending=False).head(10)

aa    833
s     598
b     592
l     590
v     408
r     251
i     203
a     143
c     121
o      82
Name: X1, dtype: int64

In [106]:
# Let's make a list of top 10 categories of X2 variable
X1_top_10 = [col for col in data['X1'].value_counts().sort_values(ascending=False).head(10).index]
X1_top_10

['aa', 's', 'b', 'l', 'v', 'r', 'i', 'a', 'c', 'o']

In [107]:
# Let's make 10 binary values for each catogory of X2 variable
for label in X1_top_10:
  data[label] = np.where(data['X1']==label, 1, 0)

data[X1_top_10].head(20)

aa  s  b  l  v  r  i  a  c  o
0    0  0  0  0  1  0  0  0  0  0
1    0  0  0  0  0  0  0  0  0  0
2    0  0  0  0  0  0  0  0  0  0
3    0  0  0  0  0  0  0  0  0  0
4    0  0  0  0  1  0  0  0  0  0
5    0  0  1  0  0  0  0  0  0  0
6    0  0  0  0  0  1  0  0  0  0
7    0  0  0  1  0  0  0  0  0  0
8    0  1  0  0  0  0  0  0  0  0
9    0  0  1  0  0  0  0  0  0  0
10   0  0  0  0  0  1  0  0  0  0
11   0  0  0  0  0  1  0  0  0  0
12   0  0  1  0  0  0  0  0  0  0
13   0  0  0  0  0  1  0  0  0  0
14   0  1  0  0  0  0  0  0  0  0
15   0  0  0  1  0  0  0  0  0  0
16   0  0  0  0  0  1  0  0  0  0
17   1  0  0  0  0  0  0  0  0  0
18   0  0  0  0  0  0  0  0  1  0
19   0  0  0  0  0  0  0  1  0  0

In [108]:
# get top 10 for each variable
X2_top_10 = [col for col in data['X2'].value_counts().sort_values(ascending=False).head(10).index]
X3_top_10 = [col for col in data['X3'].value_counts().sort_values(ascending=False).head(10).index]
X4_top_10 = [col for col in data['X4'].value_counts().sort_values(ascending=False).head(10).index]
X5_top_10 = [col for col in data['X5'].value_counts().sort_values(ascending=False).head(10).index]
X6_top_10 = [col for col in data['X6'].value_counts().sort_values(ascending=False).head(10).index]

In [109]:
# get whle set of dummy variables, for all categorical variables
    # function to create dummy variables for most frequest labels
    # we can vary the number of most frequent labels that we encode
def one_hot_top_x(df, variable, top_x_labels):
  for label in top_x_labels:
    data[variable+'_'+label] = np.where(data[variable]==label, 1, 0)

In [110]:
# Encode 'X1' variable into 10 most frequent categories
one_hot_top_x(data, 'X1', X1_top_10)
data.head()

X1  X2 X3 X4 X5 X6  aa  s  ...  X1_b  X1_l  X1_v  X1_r  X1_i  X1_a  X1_c  X1_o
0  v  at  a  d  u  j   0  0  ...     0     0     1     0     0     0     0     0
1  t  av  e  d  y  l   0  0  ...     0     0     0     0     0     0     0     0
2  w   n  c  d  x  j   0  0  ...     0     0     0     0     0     0     0     0
3  t   n  f  d  x  l   0  0  ...     0     0     0     0     0     0     0     0
4  v   n  f  d  h  d   0  0  ...     0     0     1     0     0     0     0     0

[5 rows x 26 columns]

In [111]:
# Drop extra columns
cols = ['aa',	's', 'b',	'l',	'v',	'r', 'i',	'a',	'c',	'o']
data.drop(cols, axis=1, inplace=True)

In [112]:
data.head()

X1  X2 X3 X4 X5 X6  X1_aa  ...  X1_l  X1_v  X1_r  X1_i  X1_a  X1_c  X1_o
0  v  at  a  d  u  j      0  ...     0     1     0     0     0     0     0
1  t  av  e  d  y  l      0  ...     0     0     0     0     0     0     0
2  w   n  c  d  x  j      0  ...     0     0     0     0     0     0     0
3  t   n  f  d  x  l      0  ...     0     0     0     0     0     0     0
4  v   n  f  d  h  d      0  ...     0     1     0     0     0     0     0

[5 rows x 16 columns]

In [113]:
# transform categorical variables into one hot encoding for remaining variables
one_hot_top_x(data, 'X2', X2_top_10)
one_hot_top_x(data, 'X3', X3_top_10)
one_hot_top_x(data, 'X4', X4_top_10)
one_hot_top_x(data, 'X5', X5_top_10)
one_hot_top_x(data, 'X6', X6_top_10)

In [114]:
data.head()

X1  X2 X3 X4 X5 X6  X1_aa  ...  X6_i  X6_l  X6_a  X6_h  X6_k  X6_c  X6_b
0  v  at  a  d  u  j      0  ...     0     0     0     0     0     0     0
1  t  av  e  d  y  l      0  ...     0     1     0     0     0     0     0
2  w   n  c  d  x  j      0  ...     0     0     0     0     0     0     0
3  t   n  f  d  x  l      0  ...     0     1     0     0     0     0     0
4  v   n  f  d  h  d      0  ...     0     0     0     0     0     0     0

[5 rows x 57 columns]

In [115]:
# drop intial categorical variables
cols=['X1', 'X2', 'X3', 'X4', 'X5', 'X6']
data.drop(cols, axis=1, inplace=True)
data.head()

X1_aa  X1_s  X1_b  X1_l  X1_v  X1_r  ...  X6_l  X6_a  X6_h  X6_k  X6_c  X6_b
0      0     0     0     0     1     0  ...     0     0     0     0     0     0
1      0     0     0     0     0     0  ...     1     0     0     0     0     0
2      0     0     0     0     0     0  ...     0     0     0     0     0     0
3      0     0     0     0     0     0  ...     1     0     0     0     0     0
4      0     0     0     0     1     0  ...     0     0     0     0     0     0

[5 rows x 51 columns]

In [116]:
data.shape

(4209, 51)

##### Now, We have 51 columns after one hot coding for each variable by selecting top 10 categories.